### PR TITLE
Optimize refreshing when the repo contains uncommited binary files

### PR DESCRIPTION
**Pull request by meven**

### CODE REVIEW

File: src/git/Buffer.h [hunk 1]
```diff
@@ -12,12 +12,14 @@
...
 #include "git2/buffer.h"
 
+class QByteArray;
 namespace git {
 
 class Buffer
 {
 public:
-  Buffer(const char *data, int size);
+    Buffer(const char *data, int size);
```
Comment (Philipp91) [inline src/git/Buffer.h:21]: The indentation changes in an unintended way.

File: src/git/Buffer.h [hunk 2]
```diff
@@ -12,12 +12,14 @@
 
 #include "git2/buffer.h"
 
+class QByteArray;
```
Comment (Philipp91) [inline src/git/Buffer.h:15]: Insert blank line after this.

File: src/git/Buffer.cpp [hunk 3]
```diff
@@ -8,13 +8,19 @@
...
 
 namespace git {
 
 Buffer::Buffer(const char *data, int size)
   : d(GIT_BUF_INIT_CONST(data, size))
 {}
 
+
+Buffer::Buffer(const QByteArray &data)
+  : d(GIT_BUF_INIT_CONST(data.constData(), data.size()))
```
Comment (Philipp91) [inline src/git/Buffer.cpp:21]: Might be cleaner to call the other constructor instead (don't have to repeat `GIT_BUF_INIT_CONST` and whatever people insert in the `{}` in the future).

File: src/git/Index.cpp [hunk 4]
```diff
@@ -424,10 +424,10 @@ Id Index::workdirId(const QString &path, uint32_t *mode) const
 {
   Repository repo(git_index_owner(d->index));
 
-  git_oid id;
-  if (int error = git_repository_hashfile(
-        &id, repo, path.toUtf8(), GIT_OBJECT_BLOB, nullptr))
-    return (error == GIT_EUSER) ? Id::invalidId() : Id();
+  const git_index_entry* entry = git_index_get_bypath(d->index, path.toUtf8(), GIT_INDEX_ENTRY_STAGEMASK);
```
Comment (Philipp91) [inline src/git/Index.cpp:427]: Move the star:
`const git_index_entry *entry`

Comment (Philipp91) [inline src/git/Index.cpp:427]: Are the changes in this file independent of the other changes (even if they have the same ultimate goal)? If so, please extract them to a separate commit and explain in the commit message how the changes here "work", i.e. how they make things faster (currently I don't understand).

**Review by Philipp91 — CHANGES_REQUESTED** [review]
Thanks, seems like a useful optimization!

I have a clarifying question and a few style-related nits.

**Comment by hackhaslam** [comment]
I agree that this looks like two separate changes. I'm worried that the `Index` changes break stuff. Do you remember what the problem was and why it was safe to make those changes?

File: src/git/Buffer.cpp (hunk 3 above)
Comment (meven) [reply src/git/Buffer.cpp:21]: It was a convenience, can remove.

File: src/git/Index.cpp (hunk 4 above)
Comment (meven) [reply src/git/Index.cpp]: They are independent but have the same goal : keep gitahead responding when a large not-commited file is in workdir.
Applying one and not the other will keep gitahead unresponsive in those situation when refreshing workdir, for the duration of the file read and hashing.

**Comment by meven** [comment]
> I agree that this looks like two separate changes.

One commit cannot achieve the goal without the other, see https://github.com/gitahead/gitahead/pull/601#discussion_r650357776
I have separated the two changes in two commits still, we can split the review.

> I'm worried that the `Index` changes break stuff. Do you remember what the problem was and why it was safe to make those changes?

`workdirId` is used to find in the index files that are staged or committed by path. We don't need to compute hash in those cases, we can optimize this by looking up by file path. The change does not change the behavior of the function, it returns a default `git::Id` when the file was not known in the repository, and the file id otherwise.
This change will generally make gitahead snappier as one each refresh it will spend less time computing hashes but instead reading git data.

File: src/git/Buffer.cpp (hunk 3 above)
Comment (Philipp91) [reply src/git/Buffer.cpp]: Removing is fine. FTR, what I meant was calling the first constructor from the second one.

File: src/git/Index.cpp [hunk 5]
```diff
@@ -422,15 +422,14 @@ Id Index::indexId(const QString &path, uint32_t *mode) const
 
 Id Index::workdirId(const QString &path, uint32_t *mode) const
 {
-  Repository repo(git_index_owner(d->index));
+  const git_index_entry *entry = git_index_get_bypath(d->index, path.toUtf8(), GIT_INDEX_ENTRY_STAGEMASK);
```
Comment (Philipp91) [inline src/git/Index.cpp:425]: Consider calling `Index::entry()` here, depending on whether you think that future edits/improvements to `Index::entry()` would have to be reflected here too (and could be missed by accident).

**Review by Philipp91 — COMMENTED** [review]
> The change does not change the behavior of the function, it returns a default git::Id when the file was not known in the repository, and the file id otherwise.

Previously it could also return `Id::invalidId()`. Was that a "never happens" case? Or is it preserved somehow in your new code?

Generally this seems like a good change. The new `workdirId()` is more understandable and more plausible -- previously `Index::isStaged()` would call a file hashing function, which seems quite odd.

**Comment by hackhaslam** [comment]
This definitely breaks functionality for me. The stage/unstage check boxes no longer work.

The whole point of `workdirId` is to get the hash of the contents of the file *on disk*. Getting the hash from the index misses the point. This essentially makes `workdirId` equivalent to `indexId`. In `isStaged` they always compare equal and everything is shown as always staged.

There may be other ways to short-circuit the comparison by hash of the contents. For instance, the index also stores the modification timestamp of the files. It would be interesting to compare the performance of refresh in GitAhead with the performance of `git status`. If `git status` is appreciably faster, then we can look to see what strategy they use to avoid hashing the file on disk.

**Comment by safield** [comment]
Is this commit going to be added to the fork gittyup?
